Package for Fedora

daxelrod opened the pull request:
Add specfile to build Jowl as an RPM on Fedora. This is largely based
upon the example specfile provided in
https://docs.fedoraproject.org/en-US/packaging-guidelines/Node.js/

Add packit config to allow automatic publishing of jowl into Fedora
packaging infrastructure.

daxelrod committed:
chore(ci): add Fedora rpm specfile

Add specfile to build Jowl as an RPM on Fedora. This is largely based
upon the example specfile provided in
https://docs.fedoraproject.org/en-US/packaging-guidelines/Node.js/

Since this is an executable that happens to be written in Node, rather
than a npm library, we can skip several steps that are only necessary
for libraries.

We use a tarball from GitHub as Source0 rather than the npmjs.org
published package (the latter is recommended by the packaging
guidelines) because this project considers the npmjs.org package as a
build product, not the preferred form of the source code. In addition,
the RPM needs tests and docs that aren't part of the npmjs.org package.

Manually fix shebang lines. Normally brp-mangle-shebangs should take
care of this, but it has a bug on fc35 and older where it doesn't find
node executables correctly.

daxelrod committed:
chore(ci): automate Fedora builds with Packit

Add packit config to allow automatic publishing of jowl into Fedora
packaging infrastructure.

The most awkward part of this is handling generation of the bundled node
dependency tarballs and bundled license file. Add new script
dist/packit-fix-spec-file.sh to replace packit's fix-spec-file action.

This script runs nodejs-packaging-bundler to generate the extra files.
Unfortunately that script is hardcoded to dump its output to
rpmbuild/SOURCES, so we must move it back to the dist directory where
packit expects it.

Because the tarball should be identical to one exported from GitHub, the
script does not mangle Source0 the way that Packit normally would. This
provides better documentation of the provenance of the code.

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -0,0 +1,87 @@
+%global npm_name jowl
```
Why are you using `%npm_name`? Couldn't you just use `%name`?

daxelrod replied:
Good point. I got that from a specfile that deliberately needed to use a different package name than the npm name but since they're the same here, I can just use `%name`.

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -0,0 +1,87 @@
+# Fix the shebang because brp-mangle-shebangs fails to detect this properly (rhbz#1998924)
+# This is fixed in fc36 and above
+sed -e "s|^#!/usr/bin/env node$|#!/usr/bin/node|" \
+    -i %{buildroot}%{nodejs_sitelib}/%{npm_name}/bin/jowl \
```
Isn't this a symlink?

daxelrod replied:
It's not, but to your point below, it probably should be.

daxelrod replied:
See below; making it a symlink breaks node's ability to find node_modules for reasons I don't understand. Going to leave it as-is, which is at least in line with how tape does it in Fedora.

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -0,0 +1,87 @@
+
+License:        MIT
+URL:            https://jowl.app
+Source0:        https://github.com/daxelrod/%{npm_name}/archive/refs/tags/v%{version}.tar.gz
```
You should use the URL form documented by Fedora here: https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_git_tags

daxelrod committed:
chore(ci): fix specfile to comply with guidelines

Fix specfile based upon review to comply with Fedora Packaging
Guidelines.

Remove npm_name macro. There's no need for an npm_name macro when we can
just use the name macro, since the name of the project on NPM and in
Fedora will always be the same.

Fix dangling node_modules_prod/.bin symlink. The example spec for tape
assumes that the prod node modules tarball will have a .bin directory
in it. Since Jowl's doesn't, don't try to create a symlink to that
directory, and create the directory during check so that we can combine
it with the dev node modules' .bin directory to run tests.

Use Fedora Packaging Gudidelines on Git Hosting Services to choose a
GitHub URL format for the Source0 URL. This URL is presumably more
stable, and is at least more consistent with other projects.

Add a trailing slash to file jowl directory in the files list to ensure
it is installed correctly.

Use fdupes RPM macro to replace duplicate files in node_modules with
links instead.

Add comments about how to generate the remaining sources from the
first one programmatically.

Clarify in comments why jowl/bin/jowl is not a symlink to jowl.js
(doing so messes up Node's ability to resolve the node_modules directory
correctly).

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -3,11 +3,16 @@ Version:        2.0.2
 Release:        1%{?dist}
 Summary:        CLI for JSON operations with Lodash
 
+%global gittag v%{version}
```
Just drop `%gittag` and use `v%{version}` directly in `Source0`.

daxelrod replied:
Good point. Fixed.

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -0,0 +1,90 @@
+    %{buildroot}%{nodejs_sitelib}/%{name}
+
+mkdir -p %{buildroot}%{nodejs_sitelib}/%{name}/bin
+install -p -D -m0755 src/bin/jowl.js %{buildroot}%{nodejs_sitelib}/%{name}/bin/jowl
```
Why not make this a symlink?

```bash
ln -sr %{buildroot}%{nodejs_sitelib}/%{name}/jowl.js %{buildroot}%{nodejs_sitelib}/%{name}/bin/jowl
```

daxelrod replied:
So it turns out when I try this, it messes up Node's library loading logic:

```
[root@e78b4a886621 /]# ls -al /usr/bin/jowl
lrwxrwxrwx. 1 root root 35 Sep 16 02:06 /usr/bin/jowl -> /usr/lib/node_modules/jowl/bin/jowl
[root@e78b4a886621 /]# ls -al /usr/lib/node_modules/jowl/bin/jowl 
lrwxrwxrwx. 1 root root 10 Sep 16 02:06 /usr/lib/node_modules/jowl/bin/jowl -> ../jowl.js
[root@e78b4a886621 /]# /usr/bin/jowl
...
Error: Cannot find module '../lib/format'
Require stack:
- /usr/lib/node_modules/jowl/jowl.js
```

That must be why the tape specfile I got this from did it without the symlink

Conan-Kudo replied:
That's weird, but okay...

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -0,0 +1,88 @@
+%files
+%doc docs/reference.md
+%license LICENSE %{name}-%{version}-bundled-licenses.txt
+%{nodejs_sitelib}/%{name}
```
I'd recommend adding a trailing `/` here, so that RPM unambiguously knows this is a folder.

(It's stupid that this is a thing you should do, blame POSIX)

Conan-Kudo commented on .packit.yaml:
```diff
@@ -0,0 +1,22 @@
+# add or remove files that should be synced
+files_to_sync:
+    - dist/jowl.spec
+    - .packit.yaml
```
You should also have the source generator script synced.

Conan-Kudo commented on dist/jowl.spec:
```diff
@@ -0,0 +1,88 @@
+# nodejs-packaging-bundler against Source0:
+Source1:        %{name}-%{version}-nm-prod.tgz
+Source2:        %{name}-%{version}-nm-dev.tgz
+Source3:        %{name}-%{version}-bundled-licenses.txt
```
Add the script that generates these sources as a Source so it's included in the SRPM.

daxelrod replied:
The script that generates these sources is just nodejs-packaging-bundler, which is already packaged in Fedora.  I've attempted to make the comments in packit-fix-spec-file.sh clearer that it's only necessary to run nodejs-packaging-bundler to generate sources, and the rest is all packit-specific stuff that is unecessary in a normal RPM build.

daxelrod committed:
chore(ci): clarify how to build outside of packit

It should be possible for someone to build jowl from dist-git with or
without running packit.

Ensure the packit-fix-spec-file.sh is synced to dist-git so that it can
clearly be seen by maintainers.

More clearly document that the only necessary step of
packit-fix-spec-file is running nodejs-packaging-bundler, so that it can
be easily run outside of packit.

daxelrod committed:
chore(ci): build for Rawhide as well

Copy in the default jobs that were previously implicitly being provided
by packit. Switch their targets to fedora-all in order to build against
all released Fedora versions plus Rawhide.

This package will be initially proposed for Rawhide, so it needs to
build there. In the future, we may reconsider always building against
rawhide as that may cause builds to fail for reasons outside of changes
to code.
